rust-runtime/aws-smithy-http-server/src/plugin/check_health.rs

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Middleware for handling health check requests.
+
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+
+use futures_util::Future;
+use hyper::{Body, Request, Response};
+use pin_project_lite::pin_project;
+use tower::{util::Oneshot, Layer, Service, ServiceExt};
+
+use crate::body::BoxBody;
+
+use super::Either;
+
+/// A [`tower::Layer`] used to apply [`CheckHealthService`].
+#[derive(Clone, Debug)]
+pub struct CheckHealthLayer<PingHandler> {
+    health_check_uri: &'static str,
+    ping_handler: PingHandler,
+}
+
+impl CheckHealthLayer<()> {
+    pub fn new<HandlerFuture: Future<Output = Response<BoxBody>>, H: Fn(Request<Body>) -> HandlerFuture>(
+        health_check_uri: &'static str,
+        ping_handler: H,
+    ) -> CheckHealthLayer<H> {
+        CheckHealthLayer {
+            health_check_uri,
+            ping_handler,
+        }
+    }
+}
+
+impl<S, H: Clone> Layer<S> for CheckHealthLayer<H> {
+    type Service = CheckHealthService<H, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CheckHealthService {
+            inner,
+            layer: self.clone(),
+        }
+    }
+}
+
+/// A middleware [`Service`] responsible for handling health check requests.
+#[derive(Clone, Debug)]
+pub struct CheckHealthService<H, S> {
+    inner: S,
+    layer: CheckHealthLayer<H>,
+}
+
+pin_project! {
+    /// A future that converts `F` into a compatible [`Service::Future`].
+    pub struct MappedHandlerFuture<R, S, F> {
```

Comment:
I don't think we want this to be `pub`? Should we just inline it into `CheckHealthFuture`?

Reply:
Inlined in 9fd208889ccaf8c9393640bcb02c4a755e6938f8.